src/p_keyframe.c

```diff
@@ -0,0 +1,587 @@
+
+        // Woof!
+        writep(sec->soundtarget);
+        writep(sec->thinglist);
+        writep(sec->floordata);
+        writep(sec->ceilingdata);
+        writep(sec->lightingdata);
+    }
+
+    // do lines
+    for (i = 0, li = lines; i < numlines; i++, li++)
+    {
+        write16(li->flags,
+                li->special,
+                li->tag);
+
+        write32(li->angle,
+                li->frontmusic,
+                li->backmusic);
+
+        // Woof!
+        writep(li->frontsector);
+        writep(li->backsector);
+
+        for (int j = 0; j < 2; j++)
+        {
+            if (li->sidenum[j] != NO_INDEX)
+            {
+                si = &sides[li->sidenum[j]];
+
+                write16(si->toptexture,
+                        si->bottomtexture,
+                        si->midtexture);
+
+                write32(si->textureoffset,
+                        si->rowoffset);
+            }
+        }
+    }
+}
+
+static void UnArchiveWorld(void)
+{
+    int i;
+    sector_t *sec;
+    line_t *li;
+    side_t *si;
+
+    // do sectors
+    for (i = 0, sec = sectors; i < numsectors; i++, sec++)
+    {
+        read32(&sec->floorheight,
+               &sec->ceilingheight,
+               &sec->floor_xoffs,
+               &sec->floor_yoffs,
+               &sec->ceiling_xoffs,
+               &sec->ceiling_yoffs,
+               (int32_t *)&sec->floor_rotation,
+               (int32_t *)&sec->ceiling_rotation);
+
+        read16(&sec->floorpic,
+               &sec->ceilingpic,
+               &sec->lightlevel,
+               &sec->special,
+               &sec->tag);
+
+        // Woof!
+        sec->soundtarget = readp();
+        sec->thinglist = readp();
+        sec->floordata = readp();
+        sec->ceilingdata = readp();
+        sec->lightingdata = readp();
+    }
+
+    // do lines
+    for (i = 0, li = lines; i < numlines; i++, li++)
+    {
+        read16((int16_t *)&li->flags,
+               &li->special,
+               &li->tag);
+        
+        read32((int32_t *)&li->angle,
+               &li->frontmusic,
+               &li->backmusic);
+
+        // Woof!
+        li->frontsector = readp();
+        li->backsector = readp();
+
+        for (int j = 0; j < 2; j++)
+        {
+            if (li->sidenum[j] != NO_INDEX)
+            {
+                si = &sides[li->sidenum[j]];
+              
+                read16(&si->toptexture,
+                       &si->bottomtexture,
+                       &si->midtexture);
+                
+                read32(&si->textureoffset,
+                       &si->rowoffset);
+                si->oldtextureoffset = si->textureoffset;
+                si->oldrowoffset = si->rowoffset;
+            }
+        }
+    }
+}
+
+static void ArchivePlayState(keyframe_t *keyframe)
+{
+    writex(&thinkercap, sizeof(thinkercap), 1);
+    writex(thinkerclasscap, sizeof(thinker_t), NUMTHCLASS);
+    keyframe->thinkers = M_CopyArena(thinkers_arena);
```

Comment:
Here it is, "saving with one memcpy". I have to add a lot of other stuff, but it's very light compared to the thinkers' list. This PR is still 20x faster for maps with lots of things compared to regular saves.

Reply:
I see these issues. Well, old saves could be handled by some kind of "legacy" code, i.e. the savegame restoring part of what is currently in `p_saveg.c`. But I understand that you don't want to replace this in the short term.

I was probably a bit triggered by your stating of "We will probably refactor DoLoadGame at some point" here:
https://github.com/fabiangreffrath/woof/issues/2346#issuecomment-3148683150

Reply:
> I was probably a bit triggered by your stating of "We will probably refactor DoLoadGame at some point"

Yes, it's not pretty 😄 But I like backwards-compatible saves and don't want to lose that feature.